File: lec_3/mergeSort.cpp
#include <iostream>
#include <array>
#include <cctype>
#include <queue>


using namespace std;
#define ARR_SIZE 10000

// 0 1 2 3 4

void merge(int a[], int s, int mid, int e){	
	size_t b1s = (mid - s + 1);
	size_t b2s = (e -(mid+1) +1 );
	int b1[b1s];
	int b2[b2s];

	for(int x = 0; x < b1s; x++){
		b1[x] = a[x + s];
	}
	
	for(int x = 0; x < b2s; x++){
		b2[x] = a[x+mid+1];
	}


	for(int x = 0; x < b1s; x++){
		cout << b1[x] << " ";
	}
	
	for(int x = 0; x < b2s; x++){
		cout << b2[x] << " "; 
	}

	cout << endl;
	int c = s;
	int i = 0, j = 0;

	while(i < b1s || j < b2s){
		if(i < b1s && j < b2s){
			if(b1[i] <= b2[j]){
				a[c] = b1[i];
				c++;
				i++;
			}else if(b1[i] >= b2[j]){
				a[c] = b2[j];
				c++;
				j++;
			}
		}else if(i >= b1s && j < b2s){
			a[c] = b2[j];
			c++;
			j++;
		}else if(j >= b2s && i < b1s){
			a[c] = b1[i];
			c++;
			i++;
		}
	}

}

void mergeSort(int a[], int s, int e){
	if(s < e){
		int mid = (s+e) /2;
		mergeSort(a, s, mid);
		mergeSort(a, mid+1, e);
		merge(a, s, mid, e);
	}
}





int main(void){

	int a[ARR_SIZE] = {0};
	char c;
	size_t s = 0;
	while((c = cin.get()) != EOF){
		if(isdigit(c)){
			a[s] = c - '0';
			s++;
		}
	}

 	 mergeSort(a, 0, s-1);
 	//merge(a, 0, (s/2), s-1);

	for(int i = 0; i < s; i++){
		cout << a[i] << endl;
	}
}

File: lec_3/mergesort.py
import sys
import numpy as np


def merge(a, s, mid, e):
	b1s = mid -s  + 1
	b2s = e - mid;
	b1 = []
	b2 = []
	for x in range(0,b1s):
		b1.append(a[x+s])

	for x in range(0,b2s):
		b2.append(a[x + mid+1])

	i = 0
	j = 0
	c = s
	while (i < b1s or j < b2s):
		if (i < b1s and j < b2s):
			if b1[i] <= b2[j]:
				a[c] = b1[i]
				c += 1
				i += 1
			elif b1[i] >= b2[j]:
				a[c] = b2[j]
				c += 1
				j += 1

		elif (i >= b1s and j < b2s):
			a[c] = b2[j]
			c += 1
			j += 1
			
		elif (j >= b2s and i < b1s):
			a[c] = b1[i]
			c += 1
			i += 1
			



def mergeSort(a, s, e):
	if s < e:
		mid = (s+e)/2
		mergeSort(a, s, mid)
		mergeSort(a, mid+1, e)
		merge(a, s, mid, e)



def read_file(filename):
	try:
		file = open(filename, 'r')
		d = []
		for line in file.readlines():
			d.append(int(line.split('\n')[0]))

		return d
	except IOError:
		print("Input error")
		sys.exit()



def main():
	if len(sys.argv) != 2:
		print("Usage: mergeSort.py filename.txt")
		sys.exit()

	data = read_file(sys.argv[1])
	print(data)
	ar = np.array(data)
	mergeSort(data, 0, len(data) -1)
	print(data)	
	ar = np.array(data)
	print(ar)


if __name__ == "__main__":
	main();